address_to_census_tract.py

```diff
@@ -7,81 +7,194 @@
+import re
+import requests
+import pandas as pd
 
 def main():
-    with open("data/test/testset.json", encoding = "UTF-8") as file:
+    if argv[1].endswith('.json'):
+        process_json(argv[1])  # hint: "data/test/testset.json"
+    elif argv[1].endswith('.csv'):
+        process_csv(argv[1])  # hint: "data/test/testset.csv"
+    else:
+        raise ValueError(f"Unknown file extension for file named «{argv[1]}». \
+Please choose from one of the following file extensions: \n \
+    * .csv \n \
+    * .json")
+
+def process_json(file_path):
+    with open(file_path, encoding = "UTF-8") as file:
         testset = [ json.loads(line) for line in file ]
 
-    tracts = load_geojson("data/geojsons/Washington_2016.geojson")
-
     for record in testset:
         address = record["address"]
 
         tract = None
-        latlng = address_to_latlng(address)
+        http_response = lookup_address(address)
+        standard_address = response_to_standard_address(http_response)
+        latlng = response_to_latlng(http_response)
 
         if latlng:
-            tract = latlng_to_polygon(latlng, tracts)
-
-            if not tract:
-                print(f"failed to find tract for {latlng}", file = stderr)
+            tract = latlng_to_polygon(latlng)
         else:
             print(f"failed to geocode {address}", file = stderr)
 
         # GEOID is the nationally-unique tract identifier
         result = {
+            "address": standard_address,
             "latlng": latlng,
-            "tract": tract.get("properties", {}).get("GEOID") if tract else None,
+            "tract": tract,
         }
 
         print(json.dumps({ **record, "result": result }))
 
-
-def address_to_latlng(address):
-    """Convert an address string to a list of latitude, longitude coordinates.
-
-    Currently uses Google's geocoder API for geocoding, but this could be
-    replaced with other geocoder implementations.
+def process_csv(file_path):
     """
-    return geocoder.google(address).latlng
-
+    This is going to prove difficult if addresses are separated by commas.
+    Excel may be the only way to go unless all addresses are wrapped in quotes.
+    """
+    df = pd.read_csv(file_path)
+    df['response'] = df['address'].apply(lookup_address)
+    df['standard_address'] = df['response'].apply(response_to_standard_address)
+    df['latlng'] = df['response'].apply(response_to_latlng)
+    df['final_tract'] = df['latlng'].apply(latlng_to_polygon)
+    
+    # Drop response
+    df = df.drop(columns='response', axis=1)
+    df.to_csv('data/test/out.csv') 
+
+def lookup_address(address: str) -> dict:
+    """
+    Given an address, returns a dict containing geospatial information and 
+    metadata from the SmartyStreet's geocoding API.
 
-def load_geojson(geojson_filename):
-    """Read GeoJSON file and return a list of features converted to shapes."""
+    To minimize cost, a response for each address is only looked up once.
```

Comment:
This claim in the docstring makes me think that the function is employing caching/memoization. The comment is more applicable to the _callers_ of this function, which are what control how many times an address is looked up.